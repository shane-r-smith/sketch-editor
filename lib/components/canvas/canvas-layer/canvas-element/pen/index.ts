export * from "./pen";
export * from "./pen.hooks";
