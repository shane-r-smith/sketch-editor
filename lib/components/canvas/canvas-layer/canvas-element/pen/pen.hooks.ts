import type { LineConfig } from "konva/lib/shapes/Line";
import { useMemo } from "react";
import type { CanvasElementProps } from "../canvas-element.types";

export const createPen = ({
  element,
  ...rest
}: CanvasElementProps): LineConfig => ({
  ...rest,
  ...element,
  tension: 0.5,
  lineCap: "round",
  lineJoin: "round",
  globalCompositeOperation: "source-over",
});

export const usePen = (props: CanvasElementProps): LineConfig => {
  return useMemo<LineConfig>(() => {
    return createPen(props);
  }, [props]);
};
