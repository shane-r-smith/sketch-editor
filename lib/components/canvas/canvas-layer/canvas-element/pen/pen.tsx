import { Line } from "react-konva";
import { usePen } from "./pen.hooks";
import type { CanvasElementProps } from "../canvas-element.types";

export function Pen(props: CanvasElementProps) {
  const pen = usePen(props);

  return <Line {...pen} />;
}
