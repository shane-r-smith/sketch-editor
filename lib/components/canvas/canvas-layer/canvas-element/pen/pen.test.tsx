import { render } from "@testing-library/react";
import type { CanvasElementProps } from "../canvas-element.types";
import { vi } from "vitest";
import type { LineConfig } from "konva/lib/shapes/Line";
import { Pen } from "./pen";

vi.mock("react-konva", () => ({
  Line: (props: LineConfig) => (
    <div id="mocked-line">{JSON.stringify(props, undefined, 2)}</div>
  ),
}));

describe("<Pen />", () => {
  const defaultProps: CanvasElementProps = {
    element: {
      key: "PEN",
      points: [0, 0, 100, 100],
      stroke: "#00FF00",
      strokeWidth: 10,
      strokeColourType: "HEX",
    },
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<Pen {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
