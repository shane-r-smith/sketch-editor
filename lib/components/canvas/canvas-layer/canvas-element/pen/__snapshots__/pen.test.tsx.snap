// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Pen /> > renders correctly and matches snapshot 1`] = `
<div>
  <div
    id="mocked-line"
  >
    {
  "points": [
    0,
    0,
    100,
    100
  ],
  "stroke": "#00FF00",
  "strokeWidth": 10,
  "strokeColourType": "HEX",
  "tension": 0.5,
  "lineCap": "round",
  "lineJoin": "round",
  "globalCompositeOperation": "source-over"
}
  </div>
</div>
`;
